Full package review for v0.2.0

joshwlambert opened the pull request:
This PR is to provide a platform to review the entirety of the package.

Once this review concludes I will release v0.2.0 on GitHub.

Please see the NEWS.md file for an overview of changes between v0.1.0 and v0.2.0.

This PR is unconventional as it is not intended for merging or for additional commits (unless minor) and instead comments will be converted to issues and these will be addressed in their own PRs.

CarmenTamayo commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+```
+
+```{r, generate-epidist}
+generate(ebola_si_epidist, times = 10)
````
out of curiosity where does this function come from?

joshwlambert replied:
It comes from [{distributional}](https://github.com/mitchelloharawild/distributional/blob/31ccd0c4da19d7e9334f3d0bc24b01f18117dc95/R/reexports.R#L2), but is reexported by {epiparameter} so it will be available if you use `library(epiparameter)` or `epiparameter::generate()`.

CarmenTamayo commented on vignettes/articles/data_from_epireview.Rmd:
```diff
@@ -0,0 +1,270 @@
+
+## Specifying the probability distribution if unknown
+
+There may be instances where a delay distribution is reported, but either a probability distribution is not fit to the data, or it is not reported which probability distribution the parameters correspond to. In these cases, and when a parametric probability distribution is required for a particular epidemiological task then assuming a probability distribution can be useful. 
```
Here I think it's important to clarify if you're talking about reported in epireview or in the literature because it's a bit confusing

joshwlambert replied:
Resolved in e9002b980c1687e7df6040ce0f7480103683a158.

CarmenTamayo commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+ebola_params <- ebola_data$params
+```
+
+Here we will use the serial interval for Ebola reported by @fayeChainsTransmissionControl2015a. This is stored, over two rows of the parameter table, as the mean and standard deviation, but there is no probability distribution specified. 
````
```suggestion
Here we will use the serial interval for Ebola reported by @fayeChainsTransmissionControl2015a. This is stored, over two rows of the {epireview} parameter table, as the mean and standard deviation, but there is no probability distribution specified. 
```

CarmenTamayo commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+Given that we can convert the mean and standard deviation into parameters of a probability distribution if we assume a distribution form, we can supply this data to `as_epidist()`.
+
+```{r, convert-to-epidist-assumed-prob-dist}
+ebola_si_epidist <- as_epidist(ebola_si, prob_dist = "gamma")
````
so what's going on here is that the functions to convert the summary statistics into parameters are being used?
I would add how this is being done to the description in line 248

joshwlambert replied:
Yes, if the probability distribution is assumed we can convert the mean and sd to the distribution parameters using the {epiparameter} conversion functions internally.

I'll add a description to the text.

joshwlambert replied:
Resolved in 9d2594430130da7bde930d740d5bc368369b47d5.

CarmenTamayo commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+```
+
+## Limitations
+
````
?

joshwlambert replied:
Resolved by removing the Limitations section in 7e5fbdf03af245924243536a05ea48ce1b677264. Initially I added the section header to remind myself to list some limitations to the approach, but as it is still work-in-progress functionality I'll remove for now and add some limitations once the functionality becomes more stable.

CarmenTamayo commented on vignettes/database.Rmd:
this is really nice, I rendered it in R, how will it be displayed to the users? in the list of vignettes?

joshwlambert replied:
I'm not sure I understand your question. It can be rendered locally with `vignette("database", package = "epiparameter")` or viewed on the {epiparameter} website. It is listed in the vignettes so will be findable by users using `vignette(package = "epiparameter")`.

CarmenTamayo reviewed (commented):
I reviewed the vignette `data_from_epireview.Rmd` and everything worked as expected when following along the steps, I've left some comments with questions and feedback. The main issue that I'd include in my wishlist would be to ideally find a way to simplify the process of adding bibliographic information to epidists. I think that once the par_type column is simplified, this process will be easier, and not necessary to use grepl. The process of subsetting multirow entries also would ideally be simplified for the user but I understand that it's difficult to do so given the structure of epireview. 
I really like the interface to visualise the parameters in `database.Rmd`, maybe in the future it could be expanded to have a few more relevant columns, e.g., whether entries are parameterised.

Bisaloo commented on R/coercion.R:
```diff
@@ -0,0 +1,500 @@
+    )
+  }
+  epidist <- x
+  x <- switch(func_type,
+    density = function(x) density(x = epidist, at = x),
+    cdf = function(x) epiparameter::cdf(x = epidist, q = x),
+    generate = function(x) epiparameter::generate(x = epidist, times = x),
+    quantile = function(x) quantile(x = epidist, p = x)
+  )
+
+  # return distribution function
+  x
```
```suggestion
  # return distribution function
  switch(func_type,
    density = function(x) density(x = epidist, at = x),
    cdf = function(x) epiparameter::cdf(x = epidist, q = x),
    generate = function(x) epiparameter::generate(x = epidist, times = x),
    quantile = function(x) quantile(x = epidist, p = x)
  )
```

Too many things are named x otherwise and it's a bit hard to read.

Bisaloo commented on R/coercion.R:
```diff
@@ -0,0 +1,500 @@
+  # check object as could be invalidated by user
+  validate_epidist(x)
+
+  # make data frame
+  df <- data.frame(
+    disease = x$disease,
+    pathogen = x$pathogen,
+    epi_distribution = x$epi_dist,
+    prob_distribution = I(list(prob_dist = x$prob_dist)),
+    uncertainty = I(list(uncertainty = x$uncertainty)),
+    summary_stats = I(list(summary_stats = x$summary_stats)),
+    citation = I(x$citation),
+    metadata = I(list(metadata = x$metadata)),
+    method_assess = I(list(method_assess = x$method_assess)),
+    notes = x$notes
+  )
+  row.names(df) <- NULL
+
+  # return data frame
+  df
```
```suggestion
  data.frame(
    disease = x$disease,
    pathogen = x$pathogen,
    epi_distribution = x$epi_dist,
    prob_distribution = I(list(prob_dist = x$prob_dist)),
    uncertainty = I(list(uncertainty = x$uncertainty)),
    summary_stats = I(list(summary_stats = x$summary_stats)),
    citation = I(x$citation),
    metadata = I(list(metadata = x$metadata)),
    method_assess = I(list(method_assess = x$method_assess)),
    notes = x$notes,
    row.names = NULL
  )
```

Does this not work?

Bisaloo commented on R/plot.R:
```diff
@@ -0,0 +1,127 @@
+  # capture dots
+  dots <- list(...)
+
+  oldpar <- graphics::par(no.readonly = TRUE)
+  on.exit(graphics::par(oldpar))
```
As far as I can tell, this is no longer necessary

joshwlambert replied:
I'm happy to remove these lines. 

Has there been a recent change that has led to these not being needed? If so could you share the link?

Bisaloo commented on R/epidist.R:
```diff
@@ -0,0 +1,933 @@
+    )
+    writeLines(sprintf(dist_string, family(x)))
+  } else {
+    writeLines(sprintf("Parameters: <no parameters>"))
```
```suggestion
    writeLines("Parameters: <no parameters>")
```

Bisaloo commented on R/print.R:
```diff
@@ -0,0 +1,90 @@
+        "# ", cli::symbol$info,
+        " Use `parameter_tbl()` to see a summary table of the parameters.\n",
+        "# ", cli::symbol$info, " Explore database online at: ",
+        cli::style_hyperlink(
+          text = "https://epiverse-trace.github.io/epiparameter/dev/articles/database.html", # nolint line_length_linter
+          url = "https://epiverse-trace.github.io/epiparameter/dev/articles/database.html" # nolint line_length_linter
+        )
```
```suggestion
        cli::style_hyperlink(
          text = "https://epiverse-trace.github.io/epiparameter/articles/database.html", # nolint line_length_linter
          url = "https://epiverse-trace.github.io/epiparameter/articles/database.html" # nolint line_length_linter
        )
```

After https://github.com/epiverse-trace/epiparameter/pull/338

joshwlambert replied:
Thanks for spotting!

jamesmbaazam commented on README.Rmd:
```diff
@@ -0,0 +1,137 @@
+[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11110881.svg)](https://doi.org/10.5281/zenodo.11110881)
+<!-- badges: end -->
+
+`{epiparameter}` is an `R` package that contains a library of epidemiological parameters for infectious diseases and a set classes and helper functions to be able to work with the data. It also includes functions to extract and convert parameters from reported summary statistics.
```
```suggestion
`{epiparameter}` is an `R` package that contains a library of epidemiological parameters for infectious diseases and a set of classes and helper functions to be able to work with the data. It also includes functions to extract and convert parameters from reported summary statistics.
```

jamesmbaazam commented on R/parameter_tbl.R:
```diff
@@ -0,0 +1,118 @@
+#' of publication.
+#'
+#' @inheritParams epidist
+#' @param multi_epidist Either an `<epidist>` object or a list of `<epidist>`
```
This argument is called "multiepidist" but can take one epidist. I think it could be confusing.

joshwlambert replied:
It can take one or multiple `<epidist>` objects. Here's an example where I supply all the Ebola parameters in the database.

``` r
library(epiparameter)
  ed = epidist_db(disease = "Ebola")
#> Returning 17 results that match the criteria (17 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the citation for each use the 'get_citation' function
```

``` r
  ed
#> # List of 17 <epidist> objects
#> Number of diseases: 1
#> ❯ Ebola Virus Disease
#> Number of epi distributions: 9
#> ❯ hospitalisation to death ❯ hospitalisation to discharge ❯ incubation period ❯ notification to death ❯ notification to discharge ❯ offspring distribution ❯ onset to death ❯ onset to discharge ❯ serial interval
#> [[1]]
#> Disease: Ebola Virus Disease
#> Pathogen: Ebola Virus
#> Epi Distribution: offspring distribution
#> Study: Lloyd-Smith J, Schreiber S, Kopp P, Getz W (2005). "Superspreading and
#> the effect of individual variation on disease emergence." _Nature_.
#> doi:10.1038/nature04153 <https://doi.org/10.1038/nature04153>.
#> Distribution: nbinom
#> Parameters:
#>   mean: 1.500
#>   dispersion: 5.100
#> 
#> [[2]]
#> Disease: Ebola Virus Disease
#> Pathogen: Ebola Virus-Zaire Subtype
#> Epi Distribution: incubation period
#> Study: Eichner M, Dowell S, Firese N (2011). "Incubation period of ebola
#> hemorrhagic virus subtype zaire." _Osong Public Health and Research
#> Perspectives_. doi:10.1016/j.phrp.2011.04.001
#> <https://doi.org/10.1016/j.phrp.2011.04.001>.
#> Distribution: lnorm
#> Parameters:
#>   meanlog: 2.487
#>   sdlog: 0.330
#> 
#> [[3]]
#> Disease: Ebola Virus Disease
#> Pathogen: Ebola Virus-Zaire Subtype
#> Epi Distribution: onset to death
#> Study: The Ebola Outbreak Epidemiology Team, Barry A, Ahuka-Mundeke S, Ali
#> Ahmed Y, Allarangar Y, Anoko J, Archer B, Abedi A, Bagaria J, Belizaire
#> M, Bhatia S, Bokenge T, Bruni E, Cori A, Dabire E, Diallo A, Diallo B,
#> Donnelly C, Dorigatti I, Dorji T, Waeber A, Fall I, Ferguson N,
#> FitzJohn R, Tengomo G, Formenty P, Forna A, Fortin A, Garske T,
#> Gaythorpe K, Gurry C, Hamblion E, Djingarey M, Haskew C, Hugonnet S,
#> Imai N, Impouma B, Kabongo G, Kalenga O, Kibangou E, Lee T, Lukoya C,
#> Ly O, Makiala-Mandanda S, Mamba A, Mbala-Kingebeni P, Mboussou F,
#> Mlanda T, Makuma V, Morgan O, Mulumba A, Kakoni P, Mukadi-Bamuleka D,
#> Muyembe J, Bathé N, Ndumbi Ngamala P, Ngom R, Ngoy G, Nouvellet P, Nsio
#> J, Ousman K, Peron E, Polonsky J, Ryan M, Touré A, Towner R, Tshapenda
#> G, Van De Weerdt R, Van Kerkhove M, Wendland A, Yao N, Yoti Z, Yuma E,
#> Kalambayi Kabamba G, Mwati J, Mbuy G, Lubula L, Mutombo A, Mavila O,
#> Lay Y, Kitenge E (2018). "Outbreak of Ebola virus disease in the
#> Democratic Republic of the Congo, April–May, 2018: an epidemiological
#> study." _The Lancet_. doi:10.1016/S0140-6736(18)31387-4
#> <https://doi.org/10.1016/S0140-6736%2818%2931387-4>.
#> Distribution: gamma
#> Parameters:
#>   shape: 2.400
#>   scale: 3.333
#> 
#> # ℹ 14 more elements
#> # ℹ Use `print(n = ...)` to see more elements.
#> # ℹ Use `parameter_tbl()` to see a summary table of the parameters.
#> # ℹ Explore database online at: https://epiverse-trace.github.io/epiparameter/dev/articles/database.html
```

``` r
  parameter_tbl(ed)
#> # Parameter table:
#> # A data frame:    17 × 7
#>    disease  pathogen epi_distribution prob_distribution author  year sample_size
#>    <chr>    <chr>    <chr>            <chr>             <chr>  <dbl>       <dbl>
#>  1 Ebola V… Ebola V… offspring distr… nbinom            Lloyd…  2005          13
#>  2 Ebola V… Ebola V… incubation peri… lnorm             Eichn…  2011         196
#>  3 Ebola V… Ebola V… onset to death   gamma             The E…  2018          14
#>  4 Ebola V… Ebola V… incubation peri… gamma             WHO E…  2015        1798
#>  5 Ebola V… Ebola V… incubation peri… gamma             WHO E…  2015          49
#>  6 Ebola V… Ebola V… incubation peri… gamma             WHO E…  2015         957
#>  7 Ebola V… Ebola V… incubation peri… gamma             WHO E…  2015         792
#>  8 Ebola V… Ebola V… serial interval  gamma             WHO E…  2015         305
#>  9 Ebola V… Ebola V… serial interval  gamma             WHO E…  2015          37
#> 10 Ebola V… Ebola V… serial interval  gamma             WHO E…  2015         147
#> 11 Ebola V… Ebola V… serial interval  gamma             WHO E…  2015         112
#> 12 Ebola V… Ebola V… hospitalisation… gamma             WHO E…  2015        1167
#> 13 Ebola V… Ebola V… hospitalisation… gamma             WHO E…  2015        1004
#> 14 Ebola V… Ebola V… notification to… gamma             WHO E…  2015        2536
#> 15 Ebola V… Ebola V… notification to… gamma             WHO E…  2015        1324
#> 16 Ebola V… Ebola V… onset to death   gamma             WHO E…  2015        2741
#> 17 Ebola V… Ebola V… onset to discha… gamma             WHO E…  2015        1335
```

<sup>Created on 2024-07-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

jamesmbaazam commented on vignettes/epiparameter.Rmd:
```diff
@@ -0,0 +1,203 @@
+This vignette will provide a introduction to the data stored within {epiparameter}, how to read it into R, manipulate
+the data, and the functions (and methods) implemented in the package to facilitate easy application of parameters into epidemiological pipelines.
+
+The {distributional} package is loaded as some of the {epiparameter} methods use S3 generics from this package.
```
I would suggest to cut this out. The user doesn't have to care about this.

jamesmbaazam commented on vignettes/epiparameter.Rmd:
```diff
@@ -0,0 +1,203 @@
+This vignette will provide a introduction to the data stored within {epiparameter}, how to read it into R, manipulate
+the data, and the functions (and methods) implemented in the package to facilitate easy application of parameters into epidemiological pipelines.
+
+The {distributional} package is loaded as some of the {epiparameter} methods use S3 generics from this package.
```
I would suggest to cut this out. The user doesn't have to care about this.
```suggestion

```

joshwlambert replied:
Thanks for spotting. This should have been removed after 1c52f7d8ccace4a679df911bf1b0c2c178d2036b.

jamesmbaazam commented on vignettes/epiparameter.Rmd:
```diff
@@ -0,0 +1,203 @@
+
+### Library of epidemiological parameters
+
+First, we will introduce the library, or database, of epidemiological parameters available from {epiparameter}. The library is stored in the package as a JSON file and can be read into R using the `epidist_db()` function. By default all entries in the library are supplied.
```
```suggestion
First, we will introduce the library, or database, of epidemiological parameters available from {epiparameter}. The library is stored internally and can be read into R using the `epidist_db()` function. By default all entries in the library are supplied.
```

jamesmbaazam commented on vignettes/epiparameter.Rmd:
````diff
@@ -0,0 +1,203 @@
+epi_dist_db
+```
+
+The output is a list of `<epidist>` objects, where each element in the list corresponds to an entry in the parameter database. *Technical note*: the reason the object does not look like a regular R list is because it uses a custom print method which will print a summary of the data to the console when there are more than 5 entries to prevent a large list flooding the console. When the number of d ...
+
````
The deleted text is a distraction to a user who doesn't care what's going on in the background. Maybe move it to the Design vignette?
```suggestion
The output is a list of `<epidist>` objects, where each element in the list corresponds to an entry in the parameter database. 
```

joshwlambert replied:
Happy to remove this text as you propose. There is a similar point to the technical note already in the `design_priniples.Rmd`.

jamesmbaazam commented on vignettes/epiparameter.Rmd:
````diff
@@ -0,0 +1,203 @@
+
+```{r add-to-library}
+# wrap <epidist> in list to append to database
+new_db <- append(epi_dist_db, list(covid_incubation))
````
Would be useful to see the results.

```suggestion
new_db <- append(epi_dist_db, list(covid_incubation))
new_db
```

jamesmbaazam replied:
Maybe consider a `c.epidist()` method?

jamesmbaazam commented on vignettes/epiparameter.Rmd:
````diff
@@ -0,0 +1,203 @@
+new_db <- append(epi_dist_db, list(covid_incubation))
+```
+
+Note that this only adds the parameters to the library in the environment, and does not save to the database file in the package. 
````
Here, would be useful to suggest what to do. Should they write the new db to disk?

```suggestion
Note that this only adds the parameters to the library in the environment, and does not save to the database file in the package. Hence, if you restart your R session, you will lose the changes. 
```

joshwlambert replied:
On the point of writing to disk or contributing to the {epiparameter} database this has been raised in #304. I'll commit your improvement to the text and then work on how best a user can contribute or store an expanded database in the next version.

jamesmbaazam commented on vignettes/epiparameter.Rmd:
```diff
@@ -0,0 +1,203 @@
+
+Note that this only adds the parameters to the library in the environment, and does not save to the database file in the package. 
+
+The library of epidemiological parameters is a living database, as new studies are published we hope to incorporate these. Due to the large time requirement of searching for and recording parameters in the database we welcome others to add parameters by either making a pull request to the package or adding information to the contributing spreadsheet. These will be incorporated into the database b ...
```
```suggestion
The library of epidemiological parameters is a living database, so as new studies are published we hope to incorporate these. Searching for and recording parameters in the database is extremely time-consuming, so we welcome contributions of new parameters by either making a pull request to the package or adding information to the contributing spreadsheet. These will be incorporated into the database by the package maintainers and your contributions will be acknowledged. See the [Data Collation and Synthesis Protocol](https://epiverse-trace.github.io/epiparameter/articles/data_protocol.html) vignette on information about contributing to the library of epidemiological parameters.
```

jamesmbaazam commented on vignettes/extract_convert.Rmd:
```diff
@@ -0,0 +1,275 @@
+
+### Conversion functions
+
+The conversion functions are generic functions. That means that different types of data can be supplied to the first argument and the correct operation (method) will be executed. Currently {epiparameter} provides conversion methods for `character` and `<epidist>`. If this is unclear, see the examples below for how the conversion functions can work when you supply either a `character` string or an ...
```
Once again, I think this is too technical for the general audience. All they need to use this function is that they can supply the string of the distribution name or an epidist.

```suggestion
The conversion functions can take two types of inputs as the first argument: a `character` string of the distribution or an `<epidist>` object.
```

jamesmbaazam reviewed (commented):
# `{epiparameter}` v0.1.0.9000 full package review

Thanks @joshwlambert for your hard work on this package. This review provided me an opportunity to update myself with the latest developments. I have left some line comments and will leave my long form comments here for your consideration.

## README

- It would be helpful to link the [database](https://epiverse-trace.github.io/epiparameter/articles/database.html) to the README so people can access it immediately they visit the package's page.
- You point to a googlesheet where people can contribute new parameters to the database. Could you link the data dictionary in the sheet so new contributors know what the column names mean?

## Getting Started Vignette

- The Getting Started vignette contains a lot of useful information but needs a bit of re-organising to make it accessible to the user. Currently, I fear it's a bit too technical. I would propose restructuring it as follows:
  - Introductory paragraph
  - Use case of epiparameter
  - Installing and loading epiparameter
  - The full database/library of parameters:
    - Here, introduce `epidist_db()` (with defaults) and `parameter_tbl()` (See comments about `parameter_tbl() below`).
  - Interacting with the library:
    - Here, introduce the `<epidist>` class as the single entry of the database. (In my opinion, it would be helpful to leave out a lot of the details about the class and methods and focus on "functions" since that's what most users know. This way, it will remove _distractions and barriers to understanding_ and focus on directing users on how to use the package rather than how it works in the background (the object orientation bits). The technical details can/should be moved to the Design Vignette. If you agree, I'd be happy to help with this.)
    - How to subset entries. Here, introduce `epidist_db()` with `single_epidist = TRUE` as well as the `subset` argument.
  - Working with a single entry:
       - density functions,
       - plotting function, etc.
       - extracting parameter values
       - [accessors](https://epiverse-trace.github.io/epiparameter/reference/index.html#accessors) (currently left out but important).
       -  converting or extracting parameter values from summary statistics and vice versa.
  - How to add your own entry + caveats.

## `parameter_tbl()`

- I think there is duplicated functionality between `parameter_tbl()`, the `<epidist>` `print()` method, and the [database](https://epiverse-trace.github.io/epiparameter/articles/database.html). From what I understand, `parameter_tbl()` is a way to subset the database and return a data.frame of the results. However, `epidist_db()` with `single_epidist = FALSE` returns a similar result but as a list. Moreover, it seems `parameter_tbl()` is simply a subset of the database. I think it would be helpful to have a single function that returns a consistent result of subsetting the database.

- If you are keeping `parameter_tbl()`, consider adding an `html` argument that creates an html widget for casting to the Viewer as an HTML page, similar to `stringr::str_view()`, which uses `stringr:::str_view_widget()`. This would be helpful for people who want to see the results in a more user-friendly format than what the console provides.

## Conversion and extraction functions

- Add a `@seealso` tag linking `convert_summary_stats_to_params()` and `convert_params_to_summary_stats()`.
- Consider splitting `extract_params()` into two functions:
  - `extract_params_from_range()` for `type = "range` so that it can probably be extensible to the various types of ranges (interquartile, etc), and
  - `extract_params_from_quantiles()` where type = "percentile" or "quantile".
- In the vignette [Parameter extraction and conversion in {epiparameter}](https://epiverse-trace.github.io/epiparameter/articles/extract_convert.html):
   - The use case section is really insightful but is placed too far down the vignette to motivate the functions discussed. I would suggest to start the vignette with the use case that points out the limitation in reporting unidentified distributions of percentiles and ranges (here, giving examples of studies that did that), motivate for the functionalities epiparameter provides to solve that, showcase the functionalities, then end the vignette showing how the parameters from the studies you pointed out in the beginning can be converted to a usable format.
   - The conversion examples for negative binomial all return `Inf` for the dispersion parameter. Could you provide an example where the dispersion is finite?
   - Do you need cold folding in that vignette?

## Exported functions

- Are all the exported [object checkers](https://epiverse-trace.github.io/epiparameter/reference/index.html#epi-object-checking) needed by the user or are some only for internal use?
- You probably shouldn't export the `format()` methods for `<epidist>` and `<vb_epidist>` as users are not expected to interact with it directly.

joshwlambert commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+Given that currently only delay distributions are supported for the conversion (this feature is still under active development) we will filter to only include these.
+
+```{r, subset-marburg-params}
+delay_dist_rows <- grepl(
````
Thanks for pointing out the {epireview} PR. I'll keep an eye on it and once it's merged I'll update the vignette.

joshwlambert commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+marburg_incub_epidist$metadata
+```
+
+## Creating an `<epidist>` with full citation
````
Thanks!

joshwlambert commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+The article data needs to be loaded from {epireview} using `epireview::load_epidata_raw()` rather than `epireview::load_data()` because `load_data()` subsets the bibliographic information to only provide: `"id"`, `"first_author_surname"`, `"year_publication"`, and `"article_label"` columns.
+
+```{r, load-epreview-articles}
+marburg_articles <- load_epidata_raw(
````
If more bibliographic information gets added to the `$article` table from `load_epidata()` then we can think about ingesting the article table along with the parameter table entries in `as_epidist()` to automatically create a full citation. However, we cannot do this automatically at the moment as {epireview} is not a dependency of {epiparameter} and so we cannot call `load_epidata_raw()` in {epiparameter}.

joshwlambert commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+  marburg_params[marburg_params$id %in% multi_row_ids, ]
+```
+
+This step should be verified manually to ensure that the entries that have been selected are indeed multiple rows for the same reported epidemiological parameter. We use the first two rows of this subset table, which are the mean and standard deviation for the generation time of Marburg disease. 
````
I think there can definitely be a function in the future that checks that multiple rows correspond to the same entry. It would be better if this function were part of {epireview} as it relates to their data. 

Ideally, we would move away from multi-row entries if possible and I believe that the epireview team are trying to avoid multi-row entries for all future systematic reviews.

In terms of "so that the epidist object contains all relevant parameters", this is tricky as what parameters the `<epidist>` object stores is dependent on what data from {epireview} is supplied to `as_epidist()`, for now we are dependent on the entry having some usable data (e.g. mean and SD or distribution parameters).

joshwlambert commented on vignettes/articles/data_from_epireview.Rmd:
````diff
@@ -0,0 +1,270 @@
+```{r, subset-multi-row-entries}
+multi_row_marburg_params <-
+  marburg_params[marburg_params$id %in% multi_row_ids, ]
+```
````
Resolved in 7c6f9daa933cc238a93cc14a8d44031733368581.

joshwlambert commented:
> It would be helpful to link the [database](https://epiverse-trace.github.io/epiparameter/articles/database.html) to the README so people can access it immediately they visit the package's page.

Added in be5bab305574121936263347d094916451f4852c.

> You point to a googlesheet where people can contribute new parameters to the database. Could you link the data dictionary in the sheet so new contributors know what the column names mean?

Added a link to the data dictionary in the google sheet.

> Installing and loading epiparameter

I don't think installation should be in the Get Started vignette. It's stated in the README which is a more typical place to find them.

> The Getting Started vignette contains a lot of useful information but needs a bit of re-organising to make it accessible to the user. Currently, I fear it's a bit too technical. I would propose restructuring it as follows:

Thanks for this! I've done some of the refactoring in b6b37870bc192204cf8a9a9276a51fdeeadb009f. I'll continue working to improve the documentation and make it more user friendly in the next version.

> I think there is duplicated functionality between `parameter_tbl()`, the `<epidist>` `print()` method, and the [database](https://epiverse-trace.github.io/epiparameter/articles/database.html). From what I understand, `parameter_tbl()` is a way to subset the database and return a `data.frame` of the results. However, `epidist_db()` with `single_epidist = FALSE` returns a similar result but as a list. Moreover, it seems `parameter_tbl()` is simply a subset of the database. I think it would be helpful to have a single function that returns a consistent result of subsetting the database.

You're correct in thinking that `parameter_tbl()` outputs a subset of what `epidist_db()` outputs. But I think the value add of having this function, and why I don't see them as duplicating functionality, is to offer a way to work with the epidemiological parameter data in tabular form. This allows users who are familiar with tidyverse tools to work with this data more easily and should hopefully allow them to more easily subset the list of `<epidist>` objects to select the one(s) they want to use.

In an ideal world I think we could have the `<epidist>` class appear as a `data.frame` to the user but be a list behind the scenes. IIRC this is done in the {phylobase} package (although with S4 in that case). This would then reduce the need to supplementary helper functions such as `parameter_tbl()`. If you think this would be a good direction to move in let me know and I'll give it some thought. The downside of this approach is it would be a lot of development and maintenance. 

> If you are keeping parameter_tbl(), consider adding an html argument that creates an html widget for casting to the Viewer as an HTML page, similar to stringr::str_view(), which uses stringr:::str_view_widget(). This would be helpful for people who want to see the results in a more user-friendly format than what the console provides.

I haven't used `stringr::str_view()` before, could you please raise this as an issue and explain what use case this would address. Thanks!

> Add a `@seealso` tag linking `convert_summary_stats_to_params()` and `convert_params_to_summary_stats()`.

Resolved in d74c15d50151dde2d60fb36f5c71cd712babb868.

> Consider splitting `extract_params()` into two functions:
> - `extract_params_from_range()` for `type = "range"` so that it can probably be extensible to the various types of ranges (interquartile, etc), and
> - `extract_params_from_quantiles()` where `type = "percentile"` or `"quantile"`.

The extraction functions are going to be extended in the near future (see recent suggestions on #1) I'll consider this change when I make the other changes.

> Are all the exported [object checkers](https://epiverse-trace.github.io/epiparameter/reference/index.html#epi-object-checking) needed by the user or are some only for internal use?

Usually I write them for use in other functions in {epiparameter}, but I've also thought that they are potentially handy functions for a user that wants to quickly change the contents of an `<epidist>`. I'll leave them as exported for now, if there are any that you think should be unexported let me know.

> You probably shouldn't export the `format()` methods for `<epidist>` and `<vb_epidist>` as users are not expected to interact with it directly.

Usually the `format()` methods are exported in other packages (e.g. [{dplyr}](https://tibble.tidyverse.org/reference/formatting.html)) and `format()` methods are also available in base R for several base classes, therefore I'll leave them in for now, but could be a contender to unexport if we want to slim down the NAMESPACE.

joshwlambert commented:
Thanks everyone for providing helpful comments and suggestions! I've committed a lot of the suggested changes on the `review` branch (this PR) so I'll redirect this PR to `main` and merge the changes. 

I'll then move forward with the v0.2.0 release (#339).
